qf_lib/plotting/charts/waterfall_chart.py

```diff
@@ -44,38 +64,42 @@ def _configure_axis(self):
         self.axes.set_xticks(range(len(indices) + 2))
         self.axes.set_xticklabels(['', *indices, ''])
 
+        if self.percentage:
+            self.add_decorator(AxesFormatterDecorator(y_major=PercentageFormatter(value_format=".2f")))
+
+        # calculate padding dynamically based on the difference between highest and lowest points of the chart
+        # for all parts of the charts to be visible
+        padding = (self.cumulative_sum.max() - self.cumulative_sum.min()) * 0.1
+        ymin = min(-padding, self.cumulative_sum.min() - padding)
+        ymax = max(padding, self.cumulative_sum.max() + padding)
+        self.axes.set_ylim(ymin, ymax)
+
     def _add_text(self):
         data_element_decorators = self.get_data_element_decorators()
-        self.cumulative_sum = np.cumsum(np.concatenate([d.data.values for d in data_element_decorators]))
         for index, value in enumerate([value for data_element in data_element_decorators
                                        for value in data_element.data.items()]):
-            y_loc = value[1] if index == 0 or value[0] == self.total_value else self.cumulative_sum[index]
-            text = "{:.2%}".format(value[1]) if self.percentage else value[1]
+            y_loc = value[1] if index == 0 or value[0] == self.total_label else self.cumulative_sum[index]
+            text = f"{value[1]:.2%}" if self.percentage else f"{value[1]:.2f}"
             self.add_decorator(TextDecorator(text, y=DataCoordinate(y_loc),
                                              x=DataCoordinate(index + 1),
                                              verticalalignment='bottom',
                                              horizontalalignment='center',
                                              fontsize=10))
-        self.axes.set_ylim(0, self.cumulative_sum[:-1].max() + 0.0002)
+
+    def _add_total(self):
+        series = QFSeries([self.cumulative_sum[-1]], [self.total_label])
+        self.add_decorator(DataElementDecorator(series))
 
     def _plot_waterfall(self, index, value):
-        if index == 0 or value[0] == self.total_value:
-            color = '#A6A6A6' if value[0] == self.total_value else '#4472C4' if value[1] > 0 else '#ED7D31'
+        if index == 0 or value[0] == self.total_label:
+            color = '#A6A6A6' if value[0] == self.total_label else '#4472C4' if value[1] > 0 else '#ED7D31'
```

Comment:
I would have the first if only containing the `if value[0] == self.total_label` and the merge the index == 0 in the else part, where the bottom would be `self.cumulative_sum[index - 1] if index >=1 else 0`